update documentation for custom slot mappings (#7233)
RasaHQ/rasa: docs/docs/forms.mdx

If none of the predefined [Slot Mappings](./forms.mdx#slot-mappings) fit your use
case, you can use the
[Custom Action](./actions.mdx#custom-actions) `validate_<form_name>` to write your own
extraction code. Rasa Open Source will trigger this action when the form is run.

If you're using the Rasa SDK we recommend you to extend the provided
`FormValidationAction`. When using the `FormValidationAction`, two things are needed to
extract customs slots:

> 1. Define a method `extract_<slot_name>` for every custom slot mapping which you want to
>    extract.

Reviewer: What about this:
```suggestion
1. Define a method `extract_<slot_name>` for every slot that should be extracted using a custom mapping.
```
My motivation is to get rid of the `custom slot` term as much as possible to stress out that it is the mapping that is `custom`. Though the possible confusion between `extracting` and `mapping` remains... Then perhaps this could partly do the job?
```suggestion
1. Define a method `extract_<slot_name>` for every slot that should be mapped in a custom way.
```

> 2. Optional: If a slot is not part of your mapped slots in the domain,
>    override `required_slots` to make sure the additional slot is considered by the form.

Reviewer: Wait, currently we want users to **not** include in the domain (for the given form) any slots that should use custom mappings, right? So this 2nd point should be more like two different points:
```suggestion
2. Make sure that in the domain file you list for your form only those slots that don't use custom mappings.
3. Override `required_slots` to add all slots with custom mappings to the list of slots the form should request.
```

> The following example shows the implementation of a form which extracts a slot
> `outdoor_seating` in addition to the slots which were mapped in the domain.

Reviewer: How about:
```suggestion
The following example shows the implementation of a form which extracts a slot
`outdoor_seating` in a custom way, in addition to the slots whose mappings are defined in the domain.
```
(trying to get rid of "mapped in the domain")

        self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
    ) -> Dict[Text, Any]:
        text_of_last_user_message = tracker.latest_message.get("text")
        sit_outside = "outdoor" in text_of_last_user_message

        return {"outdoor_seating": sit_outside}
```

By default the `FormValidationAction`
will automatically set the [`requested_slot`](forms.mdx#the-requested_slot-slot) to the
first slot specified in `required_slots` which is not filled.

### Dynamic Form Behavior

By default Rasa Open Source will ask for the next empty slot from the slots
listed for your form in the domain file. If you use
[custom slot mappings](forms.mdx#custom-slot-mappings) and the `FormValidationAction`,
it will ask for the first empty slot returned by the `required_slots` method. If all
slots in `required_slots` are filled the form will be be deactivated.

If needed, you can update the required slots of your form dynamically.
This is, for example, useful when you need further details based on how
a previous slot was filled or you want to change the order in which slots are requested.

If you are using the Rasa SDK, we recommend you to use the `FormValidationAction` and

> override `required_slots` to fit your dynamic behavior. Please remember to implement

Reviewer 2: ```suggestion
override `required_slots` to fit your dynamic behavior. You should implement
```

a method `extract_<slot name>` for every slot which doesn't use a predefined mapping,
as described in [Custom Slot Mappings](forms.mdx#custom-slot-mappings).
The example below will ask the user if they want to sit in

> the shade or in the sun in case they told us that they want to sit outside.

Reviewer 2: ```suggestion
the shade or in the sun in case they said they want to sit outside.
```

If none of the predefined [Slot Mappings](./forms.mdx#slot-mappings) fit your use
case, you can use the
[Custom Action](./actions.mdx#custom-actions) `validate_<form_name>` to write your own
extraction code. Rasa Open Source will trigger this action when the form is run.

If you're using the Rasa SDK we recommend you to extend the provided
`FormValidationAction`. When using the `FormValidationAction`, three steps are required
to extract customs slots:

1. Define a method `extract_<slot_name>` for every slot that should be mapped in a
   custom way.

> 2. Make sure that in the domain file you list for your form only those slots that don't
>    use custom mappings.

Reviewer: I'm thinking about using "predefined mapping" more often. @wochinge what do you think?
```suggestion
2. Make sure that in the domain file you list for your form only those slots that use 
[predefined mappings](./forms.mdx#slot-mappings).
```

Author: nice, I like it!

3. Override `required_slots` to add all slots with custom mappings to the list of slots
   the form should request.

The following example shows the implementation of a form which extracts a slot

> `outdoor_seating` in a custom way, in addition to the slots whose mappings are defined
> in the domain.

Reviewer: ```suggestion
`outdoor_seating` in a custom way, in addition to the slots which use 
[predefined mappings](./forms.mdx#slot-mappings).
```

```python
from typing import Dict, Text, List, Optional, Any

from rasa_sdk import Tracker
from rasa_sdk.executor import CollectingDispatcher
from rasa_sdk.forms import FormValidationAction

class ValidateRestaurantForm(FormValidationAction):
    def name(self) -> Text:
        return "validate_restaurant_form"

    async def required_slots(
        self,
        slots_mapped_in_domain: List[Text],
        dispatcher: "CollectingDispatcher",
        tracker: "Tracker",
        domain: "DomainDict",
    ) -> Optional[List[Text]]:
        required_slots = slots_mapped_in_domain + ["outdoor_seating"]
        return required_slots

    async def extract_outdoor_seating(
        self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
    ) -> Dict[Text, Any]:
        text_of_last_user_message = tracker.latest_message.get("text")
        sit_outside = "outdoor" in text_of_last_user_message

        return {"outdoor_seating": sit_outside}
```

By default the `FormValidationAction`
will automatically set the [`requested_slot`](forms.mdx#the-requested_slot-slot) to the
first slot specified in `required_slots` which is not filled.

### Dynamic Form Behavior

> By default Rasa Open Source will ask for the next
> [`requested_slot`](forms.mdx#the-requested_slot-slot) defined in the slot mappings
> in the domain, which is not filled yet. If you use

Reviewer: Hm, I'm trying to think of a better way to refer to the slots listed for the form in the domain. How about:
```suggestion
By default Rasa Open Source will ask for the next empty slot from the slots
listed for your form in the domain file. If you use
```

[custom slot mappings](forms.mdx#custom-slot-mappings) and the `FormValidationAction`,
it will ask for the first empty slot returned by the `required_slots` method. If all
slots in `required_slots` are filled the form will be be deactivated.

If needed, you can update the required slots of your form dynamically.
This is, for example, useful when you need further details based on how
a previous slot was filled or you want to change the order in which slots are requested.

If you are using the Rasa SDK, we recommend you to use the `FormValidationAction` and
override `required_slots` to fit your dynamic behavior. Please remember to implement

> a method `extract_<slot name>` for every slot which isn't defined in the slot mappings
> of your domain as it is described [here](forms.mdx#custom-slot-mappings).

Reviewer: ```suggestion
a method `extract_<slot name>` for every slot which doesn't use a predefined mapping,
as described in [Custom Slot Mappings](forms.mdx#custom-slot-mappings).
```

If none of the predefined [Slot Mappings](./forms.mdx#slot-mappings) fit your use
case, you can use the
[Custom Action](./actions.mdx#custom-actions) `validate_<form_name>` to write your own
extraction code. Rasa Open Source will trigger this action when the form is run.

If you're using the Rasa SDK we recommend you to extend the provided
`FormValidationAction`. When using the `FormValidationAction`, two things are needed to

> extract customs slots:

Reviewer: I think "custom slot" can lead to confusion. (It has confused me in the past.) We might want to consistently talk about "custom mapping", even though it will make some sentences more complicated...

Author: Sometimes I now just went with `additional slot` as the mapping focuses upon the extraction (not yet pushed)

This is, for example, useful when you need further details based on how
a previous slot was filled or you want to change the order in which slots are requested.

If you are using the Rasa SDK, we recommend you to use the `FormValidationAction` and
override `required_slots` to fit your dynamic behavior.
The example below will ask the user if they want to sit in
the shade or in the sun in case they told us that they want to sit outside.

```python
from typing import Text, List, Optional

from rasa_sdk.forms import FormValidationAction

class ValidateRestaurantForm(FormValidationAction):
    def name(self) -> Text:
        return "validate_restaurant_form"

    async def required_slots(
        self,
        slots_mapped_in_domain: List[Text],
        dispatcher: "CollectingDispatcher",
        tracker: "Tracker",
        domain: "DomainDict",
    ) -> Optional[List[Text]]:
        additional_slots = ["outdoor_seating"]
        if tracker.slots.get("outdoor_seating") is True:
            # If the user wants to sit outside, ask
            # if they want to sit in the shade or in the sun.
            additional_slots.append("shade_or_sun")

>         return additional_slots + slots_mapped_in_domain

Reviewer: I'm a huge fan of this <3

Reviewer: Though we should also add a note or a warning here, reminding the user to implement `extract_<slot_name>` when a new slot is added that isn't mapped in the domain.

Reviewer: I mean in the docs, not a warning in the code (there, it would be an error :-D).

If none of the predefined [Slot Mappings](./forms.mdx#slot-mappings) fit your use
case, you can use the
[Custom Action](./actions.mdx#custom-actions) `validate_{form_name}` to write your own
extraction code. Rasa Open Source will trigger this action when the form is run.

If you're using the Rasa SDK we recommend you to extend the provided
`FormValidationAction`. When using the `FormValidationAction`, two things are needed to
extract customs slots:

1. Override `required_slots`. It should return a list of slot names which your form
   requires to be filled.
2. Define a method `extract_<slot_name>` for every custom slot which you want to
   extract.

> 

Reviewer 3: I'm not sure that these two are dependent. Isn't it possible to dynamically generate the required_slot list without needing to define any `extract_<slot_name>` methods or has that changed in Rasa 2.0? It may help to make a clearer distinction between these two use cases, so users don't think they need to define these for every slot.

Author: > without needing to define any extract_<slot_name> methods 

@kearnsw Not sure I'm understanding this part. You'll always need to define an `extract_<slot name>` method for custom extractions. We could indeed dynamically generate `required_slots` but I'd rather go for explicitness than magic here. @samsucik and I iterated a bit more on the solution and updated the PR accordingly. Hope it makes more sense now.

The following example shows the implementation of a custom slot mapping which sets
the slot `outdoor_seating` based on whether the keyword `outdoor` was present in the
last user message.

```python
from typing import Dict, Text, List, Optional, Any

from rasa_sdk import Tracker
from rasa_sdk.executor import CollectingDispatcher
from rasa_sdk.forms import FormValidationAction

class ValidateRestaurantForm(FormValidationAction):
    def name(self) -> Text:
        return "validate_restaurant_form"

    async def required_slots(
        self,
        dispatcher: "CollectingDispatcher",
        tracker: "Tracker",
        domain: "DomainDict",
    ) -> Optional[List[Text]]:
        return ["outdoor_seating"]

    async def extract_outdoor_seating(
        self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
    ) -> Dict[Text, Any]:
        text_of_last_user_message = tracker.latest_message.get("text")
        sit_outside = "outdoor" in text_of_last_user_message

        return {"outdoor_seating": sit_outside}
```

By default the `FormValidationAction`
will automatically set the [`requested_slot`](forms.mdx#the-requested_slot-slot) to the

> first slot specified in `required_slots` which is not filled.

Reviewer: I think the above comment is still relevant. Could we mention what's the default behaviour, i.e. when the user doesn't override `required_slots`?

> first slot specified in `required_slots` which is not filled.

Author: > that if the user doesn't return anything from required_slots (or doesn't override required_slots at all), then the form won't request any further slots.

If `required_slots` returns `[]` it will indeed deactivate the form.

>  Could we mention what's the default behaviour, i.e. when the user doesn't override required_slots?

You meant that Rasa Open Source will then request the next slot  based on the slot mappings defined in the domain?

> first slot specified in `required_slots` which is not filled.

Author: > My understanding was that if I just want to implement custom slot mapping, I implement the corresponding extract_{slot}, nothing more.

That **only** works if this slot was mapped in the domain!

> And if I just want to change the set of required slots, I override required_slots and possibly add extract_{slot} for slots that aren't listed for this form in my domain file.

Exactly.

> first slot specified in `required_slots` which is not filled.

Reviewer: > If `required_slots` returns `[]` it will indeed deactivate the form.

Ah, that's fair now with the new way `required_slot` will be implemented in `FormValidationAction`.

> You meant that Rasa Open Source will then request the next slot based on the slot mappings defined in the domain?

Exactly.

> That only works if this slot was mapped in the domain!

Ah, right, sure. And we plan to throw a warning if the user writes a custom extractor for a slot that's already mapped in the domain, right? So, then the only recommended way to do custom mapping would be to **not** map that slot in the domain, write a custom extractor method, and add the slot to the list inside `required_slots`, right?

> first slot specified in `required_slots` which is not filled.

Author: > You meant that Rasa Open Source will then request the next slot based on the slot mappings defined in the domain?

> Exactly.

Will do 👍 

> And we plan to throw a warning if the user writes a custom extractor for a slot that's already mapped in the domain, right? 

Can do 👍 

> So, then the only recommended way to do custom mapping would be to not map that slot in the domain, write a custom extractor method, and add the slot to the list inside required_slots, right?

Sounds good 👍
